core/src/main/scala/kafka/server/AutoTopicCreationManager.scala

```diff
@@ -50,21 +52,121 @@ trait AutoTopicCreationManager {
 
   def createStreamsInternalTopics(
     topics: Map[String, CreatableTopic],
-    requestContext: RequestContext
+    requestContext: RequestContext,
+    expirationTimeMs: Long
   ): Unit
 
+  def getStreamsInternalTopicCreationErrors(
+    topicNames: Set[String],
+    currentTimeMs: Long
+  ): Map[String, String]
+
+  def close(): Unit = {}
+
+}
+
+/**
+ * Thread-safe cache that stores topic creation errors with per-entry expiration.
+ * - Expiration: maintained by a min-heap (priority queue) on expiration time
+ * - Capacity: enforced by insertion-order removal (keeps the most recently inserted entries)
+ */
+class ExpiringErrorCache(maxSize: Int) {
+
+  private case class Entry(topicName: String, errorMessage: String, expirationTimeMs: Long)
+
+  private val byTopic = new java.util.HashMap[String, Entry]()
+  private val expiryQueue = new java.util.PriorityQueue[Entry](11, new java.util.Comparator[Entry] {
+    override def compare(a: Entry, b: Entry): Int = java.lang.Long.compare(a.expirationTimeMs, b.expirationTimeMs)
+  })
+  private val lock = new ReentrantLock()
+
+  def put(topicName: String, errorMessage: String, expirationTimeMs: Long): Unit = {
+    lock.lock()
+    try {
+      val existing = byTopic.get(topicName)
+      if (existing != null) {
+        // Remove old instance from structures
+        expiryQueue.remove(existing)
+      }
+
+      val entry = Entry(topicName, errorMessage, expirationTimeMs)
+      byTopic.put(topicName, entry)
+      expiryQueue.add(entry)
+
+      // Enforce capacity by removing entries with earliest expiration time first
+      while (byTopic.size() > maxSize && !expiryQueue.isEmpty) {
+        val evicted = expiryQueue.poll()
+        if (evicted != null) {
+          val current = byTopic.get(evicted.topicName)
+          if (current != null && (current eq evicted)) {
+            byTopic.remove(evicted.topicName)
+          }
+        }
+      }
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  def getErrorsForTopics(topicNames: Set[String], currentTimeMs: Long): Map[String, String] = {
+    lock.lock()
+    try {
+      cleanupExpired(currentTimeMs)
+      val result = mutable.Map.empty[String, String]
+      topicNames.foreach { topicName =>
+        val entry = byTopic.get(topicName)
+        if (entry != null) {
+          result.put(topicName, entry.errorMessage)
+        }
+      }
+      result.toMap
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  def cleanupExpired(currentTimeMs: Long): Unit = {
+    lock.lock()
```

Comment:
Why is this not private? Isn't this called from places where we already own the lock?